An UpdateStatus state of 'rollback_completed' gives a false negative

If any service in the stack has a state of `rollback_completed`, the script assumes failure.

```diff
diff --git a/_scripts/docker-stack-wait.sh b/_scripts/docker-stack-wait.sh
index 333e847..1fee120 100755
--- a/_scripts/docker-stack-wait.sh
+++ b/_scripts/docker-stack-wait.sh
@@ -81,7 +81,7 @@ while [ "$done" != "1" ]; do
           done=2
         fi
         service_done=2
-      elif [ "$state" != "new" -a "$state" != "completed" ]; then
+      elif [ "$state" != "new" -a "$state" != "completed" -a "$state" != "rollback_completed" ]; then
         done=0
         service_done=0
       fi
```

@sudo-bmitch I agree regarding having this is a first class feature. In the meantime your script has been very useful. I ended up taking a slightly different approach with it since in our CI pipeline we already know the service names we are about to update up front:

```diff
1c1,2
< #!/bin/sh
---
> #!/bin/bash
> # shellcheck disable=SC2001,SC2086,SC2003,SC2046,SC2166,SC2220
13c14
<   echo "$(basename $0) [opts] stack_name"
---
>   echo "$(basename $0) [opts] services"
56c57
< stack_name=$1
---
> IFS=',' read -ra DEPLOY_SERVICES <<< $1
62c63
<   for service_id in $(docker stack services -q "${stack_name}"); do
---
>   for service in "${DEPLOY_SERVICES[@]}"; do
64,65c65
<     service=$(docker service inspect --format '{{.Spec.Name}}' "$service_id")
<     replicas=$(docker service ls --format '{{.Replicas}}' --filter "id=$service_id")
---
>     replicas=$(docker service ls --format '{{.Replicas}}' --filter "name=$service")
75c75
<       state=$(docker service inspect -f '{{if .UpdateStatus}}{{.UpdateStatus.State}}{{else}}new{{end}}' "$service_id")
---
>       state=$(docker service inspect -f '{{if .UpdateStatus}}{{.UpdateStatus.State}}{{else}}new{{end}}' "$service")
77c77
<       if [ "$state" = "paused" ]; then
---
>       if [ "$state" = "paused" -o "$state" = "rollback_started" -o "$state" = "rollback_completed" ]; then
99d98
<  
```

```sh
Updating service [redacted] (id: 8fky6j9chvr34wlphrq2ay2sg)
Service [redacted] is updating
Service [redacted] is replicating 2/1
Service [redacted] is completed
```

Our stack has twelve services and with this approach only the services supplied will be monitored which is typically only one. If the service fails to deploy or the healthcheck fails and a rollback is initiated, the script will also trigger a failure so our pipeline can be notified.

--- a/docker-stack-wait.sh
+++ b/docker-stack-wait.sh
@@ -5,6 +5,7 @@
 # Source repo: https://github.com/sudo-bmitch/docker-stack-wait
 
 opt_h=0
+opt_r=0
 opt_s=5
 opt_t=3600
 start_epoc=$(date +%s)
@@ -12,6 +13,7 @@ start_epoc=$(date +%s)
 usage() {
   echo "$(basename $0) [opts] stack_name"
   echo "  -h:     this help message"
+  echo "  -r:     treat a rollback as successful (by default, a rollback indicates failure)"
   echo "  -s sec: frequency to poll service state (default $opt_s sec)"
   echo "  -t sec: timeout to stop waiting"
   [ "$opt_h" = "1" ] && exit 0 || exit 1
@@ -35,14 +37,15 @@ service_state() {
   service_safe=$(echo "$service" | sed 's/[^A-Za-z0-9_]/_/g')
   state=$2
   if eval [ \"\$cache_${service_safe}\" != \"\$state\" ]; then
-    echo "Service $service is $state"
+    echo "Service $service state: $state"
     eval cache_${service_safe}=\"\$state\"
   fi
 }
 
-while getopts 'hs:t:' opt; do
+while getopts 'hrs:t:' opt; do
   case $opt in
     h) opt_h=1;;
+    r) opt_r=1;;
     s) opt_s="$OPTARG";;
     t) opt_t="$OPTARG";;
   esac
@@ -56,42 +59,68 @@ fi
 stack_name=$1
 
 # 0 = running, 1 = success, 2 = error
-done=0
-while [ "$done" != "1" ]; do
-  done=1
+stack_done=0
+while [ "$stack_done" != "1" ]; do
+  stack_done=1
   for service_id in $(docker stack services -q "${stack_name}"); do
     service_done=1
     service=$(docker service inspect --format '{{.Spec.Name}}' "$service_id")
-    replicas=$(docker service ls --format '{{.Replicas}}' --filter "id=$service_id")
-    current=$(echo "$replicas" | cut -d/ -f1)
-    target=$(echo "$replicas" | cut -d/ -f2)
-    if [ "$current" != "$target" ]; then
-      service_state "$service" "replicating $replicas"
-      done=0
-      service_done=0
-    fi
-    if [ $service_done != 0 ]; then
-      # hardcode a "new" state when UpdateStatus is not defined
-      state=$(docker service inspect -f '{{if .UpdateStatus}}{{.UpdateStatus.State}}{{else}}new{{end}}' "$service_id")
-      service_state "$service" "$state"
-      if [ "$state" = "paused" ]; then
-        # cannot complete the deployment with an update paused
-        # only change to error state when no other updates are still running
-        if [ "$done" != "0" ]; then
-          done=2
-        fi
+
+    # hardcode a "new" state when UpdateStatus is not defined
+    state=$(docker service inspect -f '{{if .UpdateStatus}}{{.UpdateStatus.State}}{{else}}new{{end}}' "$service_id")
+
+    # check for failed update states
+    case "$state" in
+      paused|rollback_paused)
         service_done=2
-      elif [ "$state" != "new" -a "$state" != "completed" ]; then
-        done=0
+        ;;
+      rollback_*)
+        if [ "$opt_r" = "0" ]; then
+          service_done=2
+        fi
+        ;;
+    esac
+
+    # identify/report current state
+    if [ "$service_done" != "2" ]; then
+      replicas=$(docker service ls --format '{{.Replicas}}' --filter "id=$service_id")
+      current=$(echo "$replicas" | cut -d/ -f1)
+      target=$(echo "$replicas" | cut -d/ -f2)
+      if [ "$current" != "$target" ]; then
+        # actively replicating service
         service_done=0
+        state="replicating $replicas"
       fi
     fi
+    service_state "$service" "$state"
+
+    # check for states that indicate an update is done
+    if [ "$service_done" = "1" ]; then
+      case "$state" in
+        new|completed|rollback_completed)
+          service_done=1
+          ;;
+        *)
+          # any other state is unknown, not necessarily finished
+          service_done=0
+          ;;
+      esac
+    fi
+
+    # update stack done state
+    if [ "$service_done" = "2" ]; then
+      # error condition
+      stack_done=2
+    elif [ "$service_done" = "0" -a "$stack_done" = "1" ]; then
+      # only go to an updating state if not in an error state
+      stack_done=0
+    fi
   done
-  if [ "$done" = "2" ]; then
-    echo "Error: Deployment will not complete"
+  if [ "$stack_done" = "2" ]; then
+    echo "Error: This deployment will not complete"
     exit 1
   fi
-  if [ "$done" != "1" ]; then
+  if [ "$stack_done" != "1" ]; then
     check_timeout
     sleep "${opt_s}"
   fi

